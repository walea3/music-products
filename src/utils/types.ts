export type Product = {
  id: number;
  title: string;
  description: string;
  artist: string;
  isFree: boolean;
  image1: string;
  image2: string;
  accentColor: string;
  backgroundColor: string;
  textColor: string;
  feature_order: string;
  price?: string;
};
