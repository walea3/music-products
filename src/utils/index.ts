export * from './product';
export * from './useAPI';
export * from './test';
