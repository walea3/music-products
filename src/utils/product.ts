export type Product = {
  id: number;
  title: string;
  description: string;
  artist: string;
  isFree: boolean;
  image1: string;
  image2: string;
  accentColor: string;
  backgroundColor: string;
  textColor: string;
  feature_order: string;
  price?: string;
};

export const BASE_PRICE = 12.99;
export const DEFAULT_BG_COLOR = '#424242';

export let naiveProductCache: { [id: number]: Product } = {};

export const hasMismatchedData = ({ id }: Product) => mismatchedDataMap[id];

export const parseProducts = (products: Product[]) => {
  naiveProductCache = products.reduce(
    (cache, product) =>
      hasMismatchedData(product) ? { ...cache, [product.id]: product } : cache,
    naiveProductCache,
  );

  return Object.keys(naiveProductCache).length
    ? products.map(fixMismatch)
    : products;
};

export const fixMismatch = (product: Product) => {
  const mismatches = mismatchedDataMap[product.id];

  return mismatches
    ? mismatches.reduce((fixedProduct, [mismatchId, ...fields]) => {
        const matchedProduct = naiveProductCache[mismatchId];
        return fields.reduce(
          (acc, field) => ({ ...acc, [field]: matchedProduct[field] }),
          fixedProduct,
        );
      }, product)
    : product;
};

const mismatchedDataMap: {
  [id: number]: [mismatchId: number, ...fields: (keyof Product)[]][];
} = {
  111: [[114, 'image1', 'image2']],
  114: [[111, 'image1', 'image2']],
};
