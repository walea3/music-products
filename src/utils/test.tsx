import { ReactElement } from 'react';
import { render } from '@testing-library/react';
import { Route, RouteProps, BrowserRouter } from 'react-router-dom';

export const renderWithRouter = (ui: ReactElement, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);

  return render(ui, { wrapper: BrowserRouter });
};

export const renderPage = (
  page: RouteProps['children'],
  path: RouteProps['path'],
  route: string,
) => renderWithRouter(<Route path={path}>{page}</Route>, { route });

// based on react-testing library default
const TEST_ID_ATTR = 'data-testid';
const TEST_ID_DELIMITER = '-';

// @TODO if desired remove test id in prod
export const testIdProp = (baseTestId: string) => (id?: string) => ({
  [TEST_ID_ATTR]: `${baseTestId}${id ? `${TEST_ID_DELIMITER}${id}` : ''}`,
});

export const getTestId = (...args: any[]) =>
  args.filter(Boolean).join(TEST_ID_DELIMITER);
