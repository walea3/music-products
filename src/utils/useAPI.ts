import { useState, useEffect, useCallback } from 'react';

export const useAPI = <T>(url: string) => {
  const [error, setError] = useState(null);
  const [data, setData] = useState<T>();
  const [isLoading, setLoading] = useState(true);

  const fetchData = useCallback(async () => {
    setLoading(true);
    setError(null);

    try {
      const res = await fetch(`/api/${url}`);
      const data: T = await res.json();

      setData(data);
    } catch (error) {
      setError(error);
    } finally {
      setLoading(false);
    }
  }, [url]);

  useEffect(() => {
    fetchData();
  }, [fetchData]);

  return [isLoading, data, error, fetchData] as const;
};
