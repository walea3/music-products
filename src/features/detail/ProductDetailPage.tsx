import { useParams } from 'react-router-dom';

import { ProductItem } from 'features/listing/components/ProductItem';
import { Product, useAPI } from 'utils';

type ProductParams = { productId: string };

export function ProductDetailPage() {
  const { productId } = useParams<ProductParams>();

  const [isLoading, product] = useAPI<Product>(`products/${productId}`);

  if (isLoading) return <div>Loading product ...</div>;
  if (!product)
    return <div>There was a problem finding product ({productId}).</div>;

  return (
    <ProductItem {...product}>
      <p>{product.description}</p>
    </ProductItem>
  );
}
