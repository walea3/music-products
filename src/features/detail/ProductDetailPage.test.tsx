import React from 'react';
import { fireEvent, RenderResult } from '@testing-library/react';

import { getTestId, useAPI, renderPage } from 'utils';

import { ProductDetailPage } from './ProductDetailPage';

jest.mock('utils/useAPI');

const useAPIMock = useAPI as jest.Mock;

describe('<ProductDetailPage />', () => {
  const mockProduct = {
    id: 123,
    title: 'Product title',
    artist: 'Artist name',
    description: 'Product description',
    isFree: true,
    image1: 'path/to/image1.jpg',
    accentColor: '#00b6b8',
    backgroundColor: '#F2F2F2',
    textColor: '#08070b',
  };

  const mockFetchProduct = jest.fn();

  const findByTestId = (wrapper: RenderResult, childTestId?: string) =>
    wrapper.getByTestId(getTestId('Products', childTestId));

  const path = '/product-:productId';
  const route = `/product-${mockProduct.id}`;

  describe('@renders', () => {
    test('when loading', () => {
      useAPIMock.mockReturnValue([true]);

      const wrapper = renderPage(<ProductDetailPage />, path, route);

      expect(wrapper.asFragment()).toMatchSnapshot();
    });

    test('with no product', () => {
      useAPIMock.mockReturnValue([false]);

      const wrapper = renderPage(<ProductDetailPage />, path, route);
      expect(wrapper.asFragment()).toMatchSnapshot();
    });

    test('with product', () => {
      useAPIMock.mockReturnValue([false, mockProduct]);

      const wrapper = renderPage(<ProductDetailPage />, path, route);
      expect(wrapper.asFragment()).toMatchSnapshot();
    });
  });
});
