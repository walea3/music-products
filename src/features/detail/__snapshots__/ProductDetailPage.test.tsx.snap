// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDetailPage /> @renders when loading 1`] = `
<DocumentFragment>
  <div>
    Loading product ...
  </div>
</DocumentFragment>
`;

exports[`<ProductDetailPage /> @renders with no product 1`] = `
<DocumentFragment>
  <div>
    There was a problem finding product (123).
  </div>
</DocumentFragment>
`;

exports[`<ProductDetailPage /> @renders with product 1`] = `
<DocumentFragment>
  .c2 {
  display: block;
  width: 100%;
  height: auto;
  margin: 0;
  -webkit-transition: -webkit-transform 0.5s ease-in-out;
  -webkit-transition: transform 0.5s ease-in-out;
  transition: transform 0.5s ease-in-out;
}

.c0 {
  position: relative;
  display: inline-block;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-top: 5px solid;
  border-color: #00b6b8;
  overflow: hidden;
  width: 90%;
  max-width: 490px;
  color: #08070b;
}

.c0:hover .c1 {
  -webkit-transform: scale(1.2);
  -ms-transform: scale(1.2);
  transform: scale(1.2);
}

.c3 {
  position: relative;
  z-index: 1;
  padding: 20px;
  background-color: #F2F2F2;
  border-bottom-left-radius: 10px;
  border-bottom-right-radius: 10px;
}

.c4 {
  margin-top: 0;
}

.c5 {
  display: inline-block;
  right: 10px;
  padding: 5px 10px;
  border-radius: 5px;
  background-color: #00b6b8;
  font-weight: bold;
  font-size: 1.2rem;
}

<article
    class="c0"
    data-testid="Product-123"
  >
    <img
      alt="Artist name"
      class="c1 c2"
      data-testid="Product-123-image"
      src="path/to/image1.jpg"
    />
    <div
      class="c3"
    >
      <h2
        class="c4"
        data-testid="Product-123-title"
      >
        Product title
      </h2>
      <p>
        Product description
      </p>
      <div
        class="c5"
        data-testid="Product-123-price"
      >
        Free
      </div>
    </div>
  </article>
</DocumentFragment>
`;
